lib/srv/discovery/fetchers/kube_services.go

```diff
@@ -0,0 +1,335 @@
+			}
+		}
+		nextToken = kubeServices.Continue
+		if nextToken == "" {
+			break
+		}
+	}
+	return result, nil
+}
+
+const (
+	protoHTTPS = "https"
+	protoHTTP  = "http"
+	protoTCP   = "tcp"
+)
+
+// Get fetches Kubernetes apps from the cluster
+func (f *KubeAppFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	kubeServices, err := f.getServices(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Converting service to apps can involve performing a HTTP ping to the service ports to determine protocol.
+	// Both services and ports inside services are processed in parallel to minimize time.
+	// We also set limit to prevent potential spike load on a cluster in case there are a lot of services.
+	g, _ := errgroup.WithContext(ctx)
+	g.SetLimit(10)
+
+	// Convert services to resources
+	var (
+		appsMu sync.Mutex
+		apps   types.Apps
+	)
+	for _, service := range kubeServices {
+		service := service
+
+		// Skip kubernetes own internal services
+		if isInternalKubeService(service) {
+			continue
+		}
+
+		// Skip service if it has type annotation and it's not 'app'
+		if v, ok := service.GetAnnotations()[types.DiscoveryTypeLabel]; ok && v != services.KubernetesMatchersApp {
+			continue
+		}
+
+		g.Go(func() error {
+			protocolAnnotation := service.GetAnnotations()[types.DiscoveryProtocolLabel]
+
+			ports, err := getServicePorts(service)
+			if err != nil {
+				f.Log.WithError(err).Errorf("could not get ports for the service %q", service.GetName())
+				return nil
+			}
+
+			portProtocols := map[v1.ServicePort]string{}
+			for _, port := range ports {
+				protocol := ""
+				switch protocolAnnotation {
+				case protoHTTPS, protoHTTP, protoTCP:
+					protocol = protocolAnnotation
+				default:
+					protocol = autoProtocolDetection(getServiceFQDN(service), port, f.ProtocolChecker)
+					if protocol == protoTCP {
+						continue
+					}
+				}
+				portProtocols[port] = protocol
+			}
+
+			for port, portProtocol := range portProtocols {
+				if len(portProtocols) == 1 {
+					port.Name = ""
+				}
+				newApp, err := services.NewApplicationFromKubeService(service, f.ClusterName, portProtocol, port)
+				if err != nil {
+					f.Log.WithError(err).Warnf("Could not get app from a Kubernetes service %q, port %d", service.GetName(), port.Port)
+					return nil
+				}
+
+				appsMu.Lock()
+				apps = append(apps, newApp)
+				appsMu.Unlock()
+			}
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return apps.AsResources(), nil
+}
+
+func (f *KubeAppFetcher) ResourceType() string {
+	return types.KindApp
+}
+
+func (f *KubeAppFetcher) Cloud() string {
+	return ""
+}
+
+func (f *KubeAppFetcher) String() string {
+	return fmt.Sprintf("KubeAppFetcher(Namespaces=%v, Labels=%v)", f.Namespaces, f.FilterLabels)
+}
+
+// autoProtocolDetection tries to determine port's protocol. It uses heuristics and port HTTP pinging if needed, provided
```

Comment:
can you describe the heuristics and what requests it tries to make?

Reply:
I've added it, though I don't think it's necessary - here code is the best comment to tell exact heuristics and it won't get stale with time.